### PR TITLE
Populating nested mutations with other relationships for @update

### LINKED ISSUE #396: Support BelongsToMany relationships in nested mutations
**Is your feature request related to a problem? Please describe.**

Currently only `BelongsTo` and `HasMany` are supported. While it is generally
possible to use two `HasMany` relationships in place of a `BelongsToMany`
relationship, this is neither practical nor intuitive.

**Describe the solution you'd like**

Provide a way to use nested mutations for [updating Many-to-Many relationships](https://laravel.com/docs/5.7/eloquent-relationships#updating-many-to-many-relationships).

### LINKED ISSUE #583: Enable updating the children of nested mutations within a single request
As discussed on slack this quick example should be possible to do:

```graphql
mutation {
  updateTermin(input: {  
    id: 1
    scheduled_at: "2019/01/01"
    customer: { 
      connect: {
        id: 1
        email: "joe@example.com" # also changed email in same request 
      }
    } 
  }) {
    id
  }
}
```

I'm always frustrated when I have to parse through my "parent" object, pick out all nested relationship, parse them into `connect`/`create` categorie for one request, to tie the relationship together. THEN I have to make another mutation, like e.g. "updateNestedChildren" to also update the data of the nested children. Ideally it should be possible to update with the same object I got from the query, without reparsing. 
The relationship is already defined in schema.graphql and/or the laravel Model.php

**Describe the solution you'd like**
I think it would be even better, if an "magical" approach would be possible, so you could just send data back and forth without reparsing it all the time. (except `__typename`, seems we always will have to omit `__typename`...)

Like this: 
```graphql
mutation {
  updateTermin(input: {  
    id: 1
    scheduled_at: "2019/01/01"
    customer: { 
        id: 1
        email: "joe@example.com" # also changed email in same request 
    } 
  }) {
    id
  }
}
```

**Describe alternatives you've considered**

I researched #131 and #207 and read the @master docs about nested mutations and asked on slack

**Additional context**
For my client I'm using `vue.js` with `vue-apollo`. I love that the querys from graphql/lighthouse is just one big json/object which I then can edit with my components. I'm frustrated that updating this is such an headache, instead of just sending it back the way it came.

From conversation over at slack @2019/01/22:

> 
> Joe [6:20 PM]
> he, I'm stil trying to figure out how to use nested mutations
> regarding this point in the docs:
> Pasted image at 2019-01-21, 6:20 PM 
> 
> I only can connect my model with existing nested relationship via the ID field, what I can't do is something like this:
> ```
> input CreateAuthorRelation {
>   connect: UpdateUserInput
>   create: CreateUserInput
> }
> 
> input UpdateUserInput {
>   id: ID
>   name: String
> }
> ```
> (edited)
> were I also update other fields, in this example, `name` at the same time.
> so what I basically want to do is to update my whole model included all nested mutations but also if something changed in the nested mutation, that should be updated at the same go. this is not possible currently, right?
> maybe a mutation example explains it best:
> ```
> mutation {
>   updatePost(input: {  
>     id: 1
>     title: "Changed Titled"
>     author: { 
>       connect: {
>         id: 1
>         description: "Also changed author descriptions"
>       }
>     } 
>   }) {
>     id
>   }
> }
> ```
> 
> Joe [6:39 PM]
> are you guys creating update/create mutation for every nested mutation? I don't get how you can work with bigger, more nested object efficently?
> 
> Joe [6:50 PM]
> do you omit every property with a nested relationship before you send it to update?
> that seems absurd?
> 
> spawnia [10:21 PM]
> Hey @Joe i think you bring up a totally valid point about updating the children of nested mutations
> This seems like it should be possible, feel free to open an issue to GitHub or provide a PR with an implementation
> We decided against allowing it in our work project, but since it can be controlled via the schema definition, Lighthouse should offer the capability for users that require such

### LINKED ISSUE #586: Enhance documentation for nested mutations
Hello,

I am new to GraphQL and tried the laravel-graphql package from Folkloreatelier. Reading the issues I found a post that guides me here and that was about nested mutations. [#313](https://github.com/Folkloreatelier/laravel-graphql/issues/313#issuecomment-421802028)

Could you please document that lighthouse feature ?

PS: I have plenty of questions and I'd like to join your Slack can you help me for that ?

### PR DESCRIPTION
**Pull request by liepaja**
- [x] Added or updated tests
- [x] Added Docs for all relevant versions

**Related Issue**

Resolves #396 
Resolves #583
Resolves #586

Related to #332

**Changes**

Adding other relationship capabilities to nested mutations when using `@update`.

| **T** indicates presence of a test

------------------------------- One to one/many

- **BelongsTo**
- [x] create **T**
- [x] update **T**
- [x] delete **T**
- [x] connect **T**
- [x] disconnect **T**

- **HasOne**
- [x] create **T**
- [x] update **T**
- [x] delete **T**

- **HasMany**
- [x] create **T**
- [x] update **T**
- [x] delete **T**

------------------------------- Many to many

- **BelongsToMany**
- [x] create **T**
- [x] update **T**
- [x] delete **T**
- [x] connect **T**
- [x] disconnect **T**
- [x] sync **T**

------------------------------- Morph one to one/many

- **MorphOne**
- [x] create **T**
- [x] update **T**
- [x] delete **T**
- [x] connect

- **MorphMany**
- [x] create **T**
- [x] update **T**
- [x] delete **T**
- [x] connect

- **MorphTo**
- [ ] no args

### CODE REVIEW

**Comment by spawnia** [comment]
Hey, can you please check the changes that are up and coming in #518 ?

Would be nice if you can base your WIP on that, since otherwise it will require another rework afterwards. Thanks

**Comment by liepaja** [comment]
@spawnia it look like a huge PR, can you point to exactly what you mean? Do you mean the return types?

**Comment by spawnia** [comment]
Take a look at the MutationExecutor and the relevant tests for it.

**Comment by spawnia** [comment]
Hey, i noticed i did not have the changes that i referred to pushed. Sorry for the confusion, you can take a look now.

**Comment by liepaja** [comment]
Ok, so basically the style

**Comment by spawnia** [comment]
@liepaja can you give me a quick update on your progress with this? A use-case has popped up in our project that requires nested BelongsToMany updates.

**Comment by lukadriel7** [comment]
Hello, thank you for the great work. How is it going?

**Comment by liepaja** [comment]
Ok, i am back on track, will try to finish it today.

**Comment by liepaja** [comment]
Core functionalities are added, will add tests and it can be merged

**Comment by JoeSchr** [comment]
that's really exciting to read, thanks!

File: src/Execution/MutationExecutor.php [hunk 1]
```diff
@@ -235,6 +249,113 @@ public static function executeUpdate(Model $model, Collection $args, ?HasMany $p
...
+                if ($operationKey === 'update') {
+                    collect($values)->each(function ($singleValues) use ($relation): void {
+                        self::executeUpdate($relation->getModel()->newInstance(), collect($singleValues), $relation);
+                    });
+                }
+
+                if ($operationKey === 'delete') {
+                    $relation->getModel()::destroy($values);
+                }
+
```
Comment (JoeSchr) [inline src/Execution/MutationExecutor.php]: Hi, just pulled your code into my project. 
Could we please also add `belongsToMany.sync()` (more details here @"Syncing Association" https://laravel.com/docs/5.7/eloquent-relationships)

```suggestion
                if ($operationKey === 'sync') {
                    $relation->sync($values);
                }
```

I put this in and it works on my end like expected

**Comment by lukadriel7** [comment]
頑張れー

**Comment by liepaja** [comment]
@JoeSchr lets leave it for another pull request, i'll try to make tests today or tomorrow so this could be finally pushed.

**Comment by spawnia** [comment]
@liepaja great progress on this, keep it up :)

**Comment by JoeSchr** [comment]
> @JoeSchr lets leave it for another pull request, i'll try to make tests today or tomorrow so this could be finally pushed.

Hi @liepaja, great work so far, but I actually think that at least `detach` is quite essential to have, otherwise you could never uncouple a many2many relationship. 
But you will probably see that once you write the tests for it, I also needed to use it first to get that something is missing.

File: src/Execution/MutationExecutor.php (hunk 1 above)
Comment (spawnia) [reply src/Execution/MutationExecutor.php]: Agree with this, will probably add this in tommorrow.

Comment (spawnia) [reply src/Execution/MutationExecutor.php]: Done

Comment (spawnia) [reply src/Execution/MutationExecutor.php]: Also, `delete` now also detaches the relation.

**Comment by spawnia** [comment]
Just a quick headsup @liepaja we are probably finishing this up by tomorrow.

**Comment by spawnia** [comment]
@liepaja this should be ready, can you give it a quick review?

**Comment by chrissm79** [comment]
Fantastic work on this one @liepaja @spawnia, glad to get this one merged in!!

File: docs/master/guides/relationships.md [hunk 2]
```diff
@@ -302,20 +304,21 @@ input UpdateAuthorRelation {
...
 
-Both `disconnect` and `delete` remove the association to the author,
-but `delete` also removes the author model itself.
+You must pass a truthy value to `disconnect` and `delete` for them to actually run.
+This structure was chosen as it is consistent with updating `BelongsToMany` relationships
+and allows the query string to be mostly static, taking a variable value to control its behaviour.
 
 ```graphql
-mutation {
+mutation UpdatePost($disconnectAuthour: Boolean){
```
[the remark this answers is not in the document]
Comment (JoeSchr) [reply docs/master/guides/relationships.md]: sorry I was on mobile and didn't see the "submit review" button when I wrote this

[the remark this answers is not in the document]
Comment (spawnia) [reply docs/master/guides/relationships.md]: Fixed in #676

**Comment by liepaja** [comment]
> @liepaja @chrissm79 this should be ready, can you give it a quick review?

Hey guys, sorry that couldn't properly finish PR. Got a tone of projects started, even a new youtube series :D. I'll take a look shortly.

**Comment by JoeSchr** [comment]
Hi,
I just tried to implement this new features for real in my project. I encounter a roadblock with a `belongsToMany` relationship, which has a pivot table (originally I'm using `withPivot(...)->as('betterName')`, but for the sake of debugging I left that one out):
```php
class Patient extends Model
{
    // many to many RELATIONSHIPS
    public function medications(): BelongsToMany
    {
        return $this->belongsToMany(Medication::class, 'patients__medications', 'patient_id', 'medication_id')->withPivot('dose_mg', 'is_rescue_medication'); //->as("details");
    }
}
```

```graphql
#schema.graphql excerpts
# Mutations {}
updateMedication(input: MedicationInput): Medication! @update(flatten: true)

# Types
# -- Medication
input MedicationInput {
    id: ID
    name: String
    pivot: MedicationDetailsInput
}

input MedicationDetailsInput {
    patient_id: ID
    medication_id: ID
    dose_mg: Float
    is_rescue_medication: Boolean
}
input MedicationRelation {
  update: [MedicationInput]
  connect: [MedicationDetailsInput]
  disconnect: [ID]
  create: [MedicationInput]
  delete: [MedicationInput]
  sync: [MedicationDetailsInput]
}
```

the error message I get is:

this seems to be the line where things are going wrong, since `pivot` property (and other name defined with `as()`) is not treated as special case:
image

sadly I don't know where to go from here, how I could resolve if there is a `pivot` defined, before passing it on into $remaining. and further how to than flatten/save it so laravel knows what to do with the `pivot` property.

but let me know if I can help you somehow to fix this.

**Comment by spawnia** [comment]
@JoeSchr there is nothing in Lighthouse right now that allows saving data to a pivot table directly through a nested BelongsToMany mutation.

You can treat your `BelongsTo` relation simply as two `BelongsTo` relationships to work around this.

If you want two allow changing pivot data as part of nested mutations, feel free to open up a new feature proposal/ PR. I am open for working that in, but the details for how the API might look are unclear to me.

**Comment by JoeSchr** [comment]
@spawnia Thanks, I worked some more on this and I think I'm almost there? 
I had two problems, one I fixed, but in a hacky way ($accessor which has the name of the pivot property, is protected) and the other (how to save the pivot property with the model relation we already have) I'm still working on.

**Comment by JoeSchr** [comment]
are there any guide(-lines) for creating a PR? I have never done it before...

**Comment by JoeSchr** [comment]
ok, I got it working, I will try to make a PR tomorrow

**Comment by spawnia** [comment]
@joeSchr https://github.com/nuwave/lighthouse/blob/master/.github/CONTRIBUTING.md

**Comment by JoeSchr** [comment]
Thanks, once I started looking for it, it was pretty easy to find and
follow ;)

**Comment by JoeSchr** [comment]
done: https://github.com/nuwave/lighthouse/pull/692

**Pull request merged**